shap/_explanation.py

```diff
@@ -613,11 +609,10 @@ def sample(self, max_samples, replace=False, random_state=0) -> Explanation:
             Random seed to use for sampling, defaults to 0.
 
         """
-        prev_seed = np.random.seed(random_state)
+        rng = np.random.RandomState(random_state)
```

Comment:
The numpy docs suggest `numpy.random.default_rng()` as the new-style way to generate random numbers. Shall we use that instead?

Nb. the numpy docs and [NEP19](https://numpy.org/neps/nep-0019-rng-policy.html#nep19) say the legacy `RandomState` is appropriate for generating data for unit tests, but I don't think this applies here.

Reply:
There's one very minor caveat that I don't see highlighted in the docs all that much, which is that `np.random.default_rng()` / `np.random.Generator` uses a different BitGenerator and SeedSequence to `RandomState`.

```python
[ins] In [1]: import numpy as np

[ins] In [2]: rng1 = np.random.default_rng(12345); print(rng1, "=>", rng1.normal())
Generator(PCG64) => -1.4238250364546312

[ins] In [3]: rng2 = np.random.RandomState(12345); print(rng2, "=>", rng2.normal())
RandomState(MT19937) => -0.20470765948471295

[ins] In [4]: rng3 = np.random.Generator(np.random.MT19937(12345)); print(rng3, "=>", rng3.normal())
Generator(MT19937) => -0.05103780855384499

[nav] In [6]: # original
         ...: prev_seed=np.random.seed(12345); print(f"original {np.random.randn() =}"); np.random.seed(prev_seed);
original np.random.randn() =-0.20470765948471295
```

As you see, only using the legacy RandomState (`rng2`) generates the exact same output as the original approach prior to this PR.
That was my main concern anyway, I recall we had a similar problem / issue previously regarding changing random generation in `shap.datasets` previously as well.

Some other refs: [NetworkX](https://networkx.org/documentation/stable/developer/nxeps/nxep-0004.html#backward-compatibility) specifically calls out this backwards incompatibility, and advocates for a major release before  this behaviour is changed as the default.

Reply:
NEP 19 (linked above) has a detailed discussion about this, which I think boils down to the fact that the reproducibility guarantee between versions has become less relevant:

> The standard practice now for bit-for-bit reproducible research is to pin all of the versions of code of your software stack, possibly down to the OS itself. The landscape for accomplishing this is much easier today than it was in 2008. We now have pip. We now have virtual machines. Those who need to reproduce simulations exactly now can (and ought to) do so by using the exact same version of numpy. We do not need to maintain stream-compatibility across numpy versions to help them.

Their recommendation is to use `default_rng` in new code, but in any case I'd be happy either way.

Reply:
Totally agree, let's use `default_rng` for new code. I'll leave it as is then (with `RandomState`), I don't see that much of a downside to keeping status quo.

Thanks for raising this regardless.